### PR TITLE
inject, install accept package spec without --spec option

### PR DESCRIPTION
**Pull request by itsayellow**
This PR allows the user to specify any valid pip package specification instead of a package name for the main argument to `pip install` and each dependency argument to `pip inject`.  It removes the `--spec` option for `pip install` because now that option is unnecessary.  And it enables a package spec for `pip inject` which formerly wasn't possible.

It takes about 2 extra seconds on my local computer to determine the package name from package spec.  If `package_spec` is a valid pypi name and not a local path, I make a shortcut and assume that it is the package name and don't try to independently determine the name (in which case it takes no extra time).

I tried to add some tests, but there could be more tests added, specifically regarding the many different VCS and other URL methods to specify pip packages.  One issue with that is that `git+...` package specifications need `git` in the path, which I wasn't sure how to do.  (https://github.com/pipxproject/pipx/pull/298#issuecomment-568642820 seems like a good way forward on this.)  Additionally, I just wasn't sure where to find good package urls to use to test.

This PR is in reference to issue #147

### CODE REVIEW

**Comment by itsayellow** [comment]
@cs01 feel free to modify code if it seems fun to you.  Otherwise I'm happy to do the normal review process...

**Comment by itsayellow** [comment]
Note, I may be out of communication for the next few days, but I may check in occasionally.

**Comment by itsayellow** [comment]
Oh and before I forget, I was also thinking about making the `Venv` member variable `_python` public, meaning just changing its name to `python`.  Because I use it in commands.py in this PR.

**Comment by cs01** [comment]
The _python name change is fine with me.

**Comment by clbarnes** [comment]
For compatibility, `_python` could be made a property which throws a DeprecationWarning but still returns the right thing, then it can be removed in a couple of release's time.

**Comment by itsayellow** [comment]
> For compatibility in the extremely unlikely event that anyone else is using it (was it in any previous releases?), `_python` could be made a property which throws a DeprecationWarning but still returns the right thing, then it can be removed in a couple of release's time.

AFAIK it was just an internal variable inside of Venv (hence the underscore).  I assume that the Venv author just assumed it would never be used outside.

I'm guessing that nobody was using it, and they were already "warned" when it had the starting underscore that they shouldn't.  😄

File: src/pipx/commands/commands.py [hunk 1]
```diff
@@ -60,18 +70,47 @@ def install(
...
+            include_dependencies,
+            force=force,
         )
     except (Exception, KeyboardInterrupt):
         print("")
         venv.remove_venv()
         raise
 
 
+def _package_name_from_spec(package_spec: str, python: str) -> str:
```
[the remark this answers is not in the document]
Comment (itsayellow) [reply src/pipx/commands/commands.py]: Yes!  I used to use the normal `Venv.install_package()` which automatically did the animation, but when I refactored a bit I didn't realize I lost animation.

**Comment by itsayellow** [comment]
Regarding the animation:
* I unquoted the package name, because other non-animated pipx messages refer to the package name unquoted and I was trying to stay consistent with them.  (e.g. black not 'black')
* The package specification __is__ quoted, because that made a lot of sense to me.
* I made a `util.py` function to return a string of the package name plus specification (if any) so that the format of this could be standardized across multiple different messages.

**Comment by itsayellow** [comment]
I added your archived black zip-file to the list of tested install packages.  In the process I discovered and fixed a bug in `abspath_if_local()`.

**Comment by itsayellow** [comment]
I also note (and not sure if this is a future PR or this one) that  `pipx list` currently does not mention a package specification and it would be helpful if they did.

**Comment by itsayellow** [comment]
I think I addressed all of your bullet-point TODOs, ready for another review.

**Comment by itsayellow** [comment]
I earmarked a testing TODO with newly-created Issue #303 that would be helpful to test `git+https://...` package specifications.

File: src/pipx/venv.py [hunk 2]
```diff
@@ -212,6 +215,25 @@ def install_package(
         if self.package_metadata[package].package_version is None:
             raise PackageInstallFailureError
 
+    def install_package_no_deps(
+        self, package_or_url: str, pip_args: List[str]
+    ) -> Optional[str]:
+        package: Optional[str]
+
+        with animate(f"resolving {package_or_url!r}", self.do_animation):
```
Comment (cs01) [inline src/pipx/venv.py]: `resolving` seems ambiguous to me. What do you think of something more descriptive like

* `determining package name from {...}`
* `resolving package name from {...}`
* `getting package name from {...}`

File: src/pipx/venv.py [hunk 3]
```diff
@@ -292,11 +314,11 @@ def run_app(self, app: str, app_args: List[str]) -> int:
...
             return 130  # shell code for Ctrl-C
 
-    def _upgrade_package_no_metadata(
-        self, package_or_url: str, pip_args: List[str]
-    ) -> None:
-        with animate(f"upgrading package {package_or_url!r}", self.do_animation):
-            self._run_pip(["install"] + pip_args + ["--upgrade", package_or_url])
+    def _upgrade_package_no_metadata(self, package: str, pip_args: List[str]) -> None:
+        with animate(
+            "upgrading " + full_package_description(package, package), self.do_animation
```
Comment (cs01) [inline src/pipx/venv.py]: nit: can f-strings be used instead of adding strings together (applies elsewhere in this PR as well)?

File: src/pipx/util.py [hunk 4]
```diff
@@ -127,3 +128,17 @@ def run(cmd: Sequence[Union[str, Path]], check=True) -> int:
...
+
+def valid_pypi_name(package_name: str) -> bool:
+    # https://www.python.org/dev/peps/pep-0508/#names
+    return bool(
+        re.search(r"^([A-Z0-9]|[A-Z0-9][A-Z0-9._-]*[A-Z0-9])$", package_name, re.I)
+    )
+
+
+def full_package_description(package, package_spec):
+    name_str = f"{package}"
```
Comment (cs01) [inline src/pipx/util.py]: Why is this necessary?

File: src/pipx/util.py [hunk 5]
```diff
@@ -127,3 +128,17 @@ def run(cmd: Sequence[Union[str, Path]], check=True) -> int:
...
+    if package != package_spec:
+        name_str += f" from spec. {package_spec!r}"
```
Comment (cs01) [inline src/pipx/util.py]: Remove `.` after `spec`

File: src/pipx/venv.py [hunk 6]
```diff
@@ -173,21 +176,21 @@ def install_package(
...
+            pip_args = []
+        if package is None:
+            # If no package name is supplied, install only main package
+            #   first in order to see what its name is
+            package = self.install_package_no_deps(package_or_url, pip_args)
+
+        install_message = f"installing {package}"
+        if package != package_or_url:
+            install_message += f" from specification {package_or_url!r}"
+
```
Comment (cs01) [inline src/pipx/venv.py]: Looks like this can be replaced with a call to `full_package_description()`?

File: docs/examples.md [hunk 7]
```diff
@@ -27,10 +27,10 @@ pipx --version  # prints pipx version
...
 pipx --python pythonX pycowsay
 pipx --spec pycowsay==2.0 pycowsay --version
-pipx --spec git+https://github.com/ambv/black.git black
-pipx --spec git+https://github.com/ambv/black.git@branch-name black
-pipx --spec git+https://github.com/ambv/black.git@git-hash black
-pipx --spec https://github.com/ambv/black/archive/18.9b0.zip black --help
+pipx --spec git+https://github.com/psf/black.git black
+pipx --spec git+https://github.com/psf/black.git@branch-name black
+pipx --spec git+https://github.com/psf/black.git@git-hash black
+pipx --spec https://github.com/psf/black/archive/18.9b0.zip black --help
```
Comment (cs01) [inline docs/examples.md]: Not introduced in this PR but I just noticed these are all missing the `run` argument.

**Comment by itsayellow** [comment]
> Just want to confirm that we are leaving the `run` command alone in this PR? i.e. it still has the optional `--spec` argument?

Confirmed, I left `run` alone.  Because the only required argument is app/package_name, the extra `--spec` option still makes sense as an additional option.

File: src/pipx/util.py (hunk 4 above)
Comment (itsayellow) [reply src/pipx/util.py]: Ah you're right, I don't need the f-string if that's what you're referring to.  It fell out in the refactoring that way.

File: src/pipx/venv.py (hunk 6 above)
Comment (itsayellow) [reply src/pipx/venv.py]: Oops, actually those three lines can be deleted, I replaced install_message with `full_package_description()` in animate later on!

File: src/pipx/util.py (hunk 5 above)
Comment (itsayellow) [reply src/pipx/util.py]: Done.

File: src/pipx/util.py (hunk 4 above)
Comment (itsayellow) [reply src/pipx/util.py]: I simplified the function full_package_description and removed this.

File: docs/examples.md (hunk 7 above)
Comment (itsayellow) [reply docs/examples.md]: True, and earlier examples in this block do not actually work (it won't work to put run options before `run`.) 

I'll fix them too.

File: src/pipx/venv.py (hunk 2 above)
Comment (itsayellow) [reply src/pipx/venv.py]: Yes, and "resolving" might be a little more loaded, it's probably too associated with package resolution.

Maybe determining package name?

File: src/pipx/venv.py (hunk 3 above)
Comment (itsayellow) [reply src/pipx/venv.py]: Done.

File: src/pipx/venv.py (hunk 2 above)
Comment (itsayellow) [reply src/pipx/venv.py]: Done.  Changed to "determining package name from "

File: src/pipx/venv.py [hunk 8]
```diff
@@ -173,21 +176,20 @@ def install_package(
...
-        with animate(f"installing package {package_or_url!r}", self.do_animation):
-            if pip_args is None:
-                pip_args = []
+        if pip_args is None:
+            pip_args = []
+        if package is None:
+            # If no package name is supplied, install only main package
+            #   first in order to see what its name is
+            package = self.install_package_no_deps(package_or_url, pip_args)
+            # TODO 20191228: should we just fail if we cannot determine package?
```
Comment (itsayellow) [inline src/pipx/venv.py]: I have a TODO here we should resolve.  I'm thinking if we cannot determine package name here we should just fail at this point before we try to install anything else.  Does that sound good?

Comment (itsayellow) [reply src/pipx/venv.py]: Not sure if I need to start a review or not (github mechanics) but mostly I'm asking @cs01 and any other interested parties.

**Comment by itsayellow** [comment]
I made install_package error out if package name cannot be determined.  I think this should be the wisest thing to do.

File: src/pipx/venv.py [hunk 9]
```diff
@@ -173,32 +176,25 @@ def install_package(
...
-                installed_packages = self.list_installed_packages() - old_package_set
-                if len(installed_packages) == 1:
-                    package = installed_packages.pop()
-                    logging.info(f"Determined package name: '{package}'")
-                else:
-                    package = None
+                logging.warning(
+                    f"Cannot determine package name for spec {package_or_url!r}. "
+                )
+                raise PackageInstallFailureError
```
Comment (cs01) [inline src/pipx/venv.py]: I agree an error should be raised if the package name cannot be determined. However, when a `PackageInstallFailureError` is raised, we could propagate more specific failure information to the user instead of unconditionally printing this message:
```
        except PackageInstallFailureError:
            venv.remove_venv()
            raise PipxError(
                f"Could not install package {package_name}. Is the name or spec correct?"
            )
```

I think it would be better if we could raise `PackageInstallFailureError` with specific failure text, and print that text directly to the user. i.e. `raise PackageInstallFailureError( f"Cannot determine package name for spec {package_or_url!r}.")` and then in `main.py` do

```
    except (PipxError, PackageInstallFailureError) as e:
```

File: src/pipx/commands/commands.py [hunk 10]
```diff
@@ -60,18 +70,47 @@ def install(
...
+        return package_name
+
+    with tempfile.TemporaryDirectory() as temp_venv_dir:
+        venv = Venv(Path(temp_venv_dir), python=python)
+        venv.create_venv(venv_args=[], pip_args=[])
+        package_name = venv.install_package_no_deps(
+            package_or_url=package_spec, pip_args=[]
+        )
+        if package_name is None:
+            raise PipxError(f"Unable to validate {package_spec}")
```
Comment (cs01) [inline src/pipx/commands/commands.py]: I just ran into this locally while testing. 
```
>> pipx install gdbgui
Unable to validate gdbgui
```
This error message is technically correct, but I think a little more context and next steps would be helpful. Maybe something like
```
raise PipxError(f"Package was found and installed, but it is invalid. Verify the package  {package_spec} can be installed with pip and then try again.
```

Comment (itsayellow) [reply src/pipx/commands/commands.py]: I'm going to have to investigate that, because it shouldn't be getting to that error message because gdbgui is a valid pypi name and presumably not a local path for your.

**Comment by itsayellow** [comment]
I revamped the package errors.  I subclassed `PackageInstallationFailureError` from `PipxError`.  Now I'm not sure if there's a value to having the separate `PackageInstallationFailureError`?

Currently if `pip install` subprocess fails, it will blab all of the pip errors to stderr.  It's a little raw, but I think useful, so I left it.  I could try and intercept this and send it to `logging.info()` if we wanted to clean things up a bit.

**Comment by itsayellow** [comment]
I removed the specific `PackageInstallFailureError` because we were never specifically using it instead of the more general `PipxError`.

File: src/pipx/commands/commands.py [hunk 11]
```diff
@@ -48,30 +58,51 @@ def install(
...
     except (Exception, KeyboardInterrupt):
         print("")
         venv.remove_venv()
         raise
 
 
+def _package_name_from_spec(package_spec: str, python: str) -> str:
+    start_time = time.time()
+
+    package_name: Optional[str]
```
Comment (cs01) [inline src/pipx/commands/commands.py]: I believe this is not Optional?

Comment (itsayellow) [reply src/pipx/commands/commands.py]: Yes!  thanks for pointing it out, I had meant to delete that.
